### PR TITLE
Introduce a ViewModel to DribbbleShot and move logic.

### LINKED ISSUE #382: [DR] Introduce ViewModel to DribbbleShot

### PR DESCRIPTION
**Pull request by nickbutcher**
Fixes #382

### CODE REVIEW

**Comment by nickbutcher** [comment]
This PR includes a copy of `TestUtils` which shouldn't be needed once we land #418 but otherwise is ready for review.

File: core/src/main/java/io/plaidapp/core/util/customtabs/CustomTabActivityHelper.java [hunk 1]
```diff
@@ -36,9 +36,9 @@
  */
 public class CustomTabActivityHelper {
     private CustomTabsSession mCustomTabsSession;
-    private CustomTabsClient mClient;
+    CustomTabsClient mClient;
```
[the remark this answers is not in the document]
Comment (nickbutcher) [reply core/src/main/java/io/plaidapp/core/util/customtabs/CustomTabActivityHelper.java]: Nope, this is to eliminate a synthetic accessor.

File: dribbble/src/main/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import io.plaidapp.core.dribbble.data.api.model.Shot
+
+/**
+ * A UseCase which prepares the information required to share a shot.
+ */
+class GetShareShotInfoUseCase(private val imageUriProvider: DribbbleImageUriProvider) {
+
+    @WorkerThread
+    operator fun invoke(shot: Shot): ShareShotInfo {
+        val url = shot.images.best() ?: throw IllegalArgumentException()
```
Comment (florina-muntenescu) [inline dribbble/src/main/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCase.kt]: you're not testing the throw case.

File: dribbble/src/test/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        // Given a shot with a png image
+        val shot = withUrl("https://cdn.dribbble.com/users/6295/screenshots/2344334/plaid_dribbble.png")
+
+        // When invoking the use case
+        val shareInfo = getShareShotInfoUseCase(shot)
+
+        // Then the expected share info is returned
+        assertNotNull(shareInfo)
+        assertEquals(shot.title, shareInfo.title)
+        assertFalse(shareInfo.shareText.isNullOrBlank())
```
Comment (florina-muntenescu) [inline dribbble/src/test/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCaseTest.kt]: should we test that this contains the title, user name and url?

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import io.plaidapp.core.data.CoroutinesContextProvider
+import io.plaidapp.dribbble.provideGetShareShotInfoUseCase
+import io.plaidapp.dribbble.ui.shot.DribbbleShotViewModel
+
+/**
+ * Factory for Dribbble [ViewModel]s
+ */
+class DribbbleViewModelFactory(
+    private val contextProvider: CoroutinesContextProvider,
+    private val context: Context
```
Comment (florina-muntenescu) [inline dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt]: I prefer passing the use case, so the factory doesn't need to hold a reference to the context.

File: dribbble/src/test/java/io/plaidapp/dribbble/TestData.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import io.plaidapp.core.dribbble.data.api.model.Shot
+import io.plaidapp.core.dribbble.data.api.model.User
+
+/**
+ * Dribbble test data
+ */
+
+val player = User(
+    id = 1L,
+    name = "Nick Butcher",
```
Comment (florina-muntenescu) [inline dribbble/src/test/java/io/plaidapp/dribbble/TestData.kt]: Nick, always a player...

File: dribbble/src/test/java/io/plaidapp/dribbble/TestUtils.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * Helper functions that are workarounds to kotlin Runtime Exceptions when using kotlin.
+ */
+
+/**
+ * Returns Mockito.any() as nullable type to avoid java.lang.IllegalStateException when
+ * null is returned.
+ */
+
+fun <T> any(): T {
```
Comment (florina-muntenescu) [inline dribbble/src/test/java/io/plaidapp/dribbble/TestUtils.kt]: once we have Mockito-Kt we should get rid of these

File: core/src/main/java/io/plaidapp/core/util/event/EventObserver.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * An [Observer] for [Event]s, simplifying the pattern of checking if the [Event]'s content has
+ * already been consumed.
+ *
+ * [onEventUnconsumedContent] is *only* called if the [Event]'s contents has not been consumed.
+ */
+class EventObserver<T>(private val onEventUnconsumedContent: (T) -> Unit) : Observer<Event<T>> {
+    override fun onChanged(event: Event<T>?) {
+        event?.consume()?.let { value ->
```
Comment (XinyueZ) [inline core/src/main/java/io/plaidapp/core/util/event/EventObserver.kt]: save into one line :         `event?.consume()?.let (onEventUnconsumedContent)`

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+) : ViewModelProvider.Factory {
+
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
```
Comment (XinyueZ) [inline dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt]: @Suppress("UNCHECKED_CAST") for `as T`

File: dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import java.io.File
+
+/**
+ * A class responsible for resolving an image as identified by Url into a sharable [Uri].
+ */
+class DribbbleImageUriProvider(private val context: Context) {
+
+    init {
+        // Only accept application context to avoid leaks
+        if (context.applicationContext != context) throw IllegalArgumentException()
```
Comment (XinyueZ) [inline dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt]: When the  `DribbbleImageUriProvider` really need `Application` as mandatory `Context`, I prefer *Change*  of constructor with param of `Application`.

After that, we can save lines 33-36 and 100% programming with contract safe.

File: dribbble/src/main/java/io/plaidapp/dribbble/Injection.kt [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * Once we have a dependency injection framework or a service locator, this should be removed.
+ */
+
+fun provideViewModelFactory(context: Context) =
+    DribbbleViewModelFactory(provideCoroutinesContextProvider(), context)
+
+fun provideGetShareShotInfoUseCase(context: Context) =
+    GetShareShotInfoUseCase(provideDribbbleImageUriProvider(context))
+
+fun provideDribbbleImageUriProvider(context: Context) =
```
Comment (XinyueZ) [inline dribbble/src/main/java/io/plaidapp/dribbble/Injection.kt]: The `DribbbleImageUriProvider` needs `Application` really. I prefer to use `application: Application`  as param. The programming with contract is easy for shrink code lines.

File: dribbble/src/main/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCase.kt (hunk 2 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/domain/GetShareShotInfoUseCase.kt]: done

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt (hunk 4 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/ui/DribbbleViewModelFactory.kt]: done.

File: dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt (hunk 9 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt]: done.

File: dribbble/src/main/java/io/plaidapp/dribbble/Injection.kt (hunk 10 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/Injection.kt]: done.

File: dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt (hunk 9 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/domain/DribbbleImageUriProvider.kt]: Actually the same can be achieved by passing in a `Context` but only holding on to the application context. This feels less restrictive on downstream consumers.

File: dribbble/build.gradle [hunk 11]
```diff
@@ -42,4 +42,16 @@ repositories {
 
 dependencies {
     implementation project(':app')
+
+    testImplementation project(':test_shared')
+    testImplementation "junit:junit:${versions.junit}"
```
Comment (keyboardsurfer) [inline dribbble/build.gradle]: nit: Please sort the dependencies

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/DribbbleShotActivity.kt [hunk 12]
```diff
@@ -266,21 +278,28 @@ class DribbbleShotActivity : AppCompatActivity() {
...
                 )
                 .addDefaultShareMenuItem()
                 .build(),
-            url?.toUri()
+            url
         )
     }
 
+    private fun shareShot(shareInfo: ShareShotInfo) {
+        ShareCompat.IntentBuilder.from(this)
```
Comment (keyboardsurfer) [inline dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/DribbbleShotActivity.kt]: `with(shareInfo)` makes this less verbose.

File: dribbble/build.gradle (hunk 11 above)
Comment (nickbutcher) [reply dribbble/build.gradle]: done

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/DribbbleShotActivity.kt (hunk 12 above)
Comment (nickbutcher) [reply dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/DribbbleShotActivity.kt]: done

**Pull request merged**